packages/next/client/page-loader.js

```diff
@@ -76,6 +91,60 @@ export default class PageLoader {
     )
   }
 
+  /**
+   * @param {string} href the route href (file-system path)
+   * @param {string} asPath the URL as shown in browser (virtual path); used for dynamic routes
+   */
+  prefetchData(href, asPath) {
+    const { pathname: hrefPathname, query } = parse(href, true)
+    const { pathname: asPathname } = parse(asPath)
+
+    const route = normalizeRoute(hrefPathname)
+    return this.promisedSsgManifest.then(
+      (
+        s,
+        _dataHref,
+        _dynamic_regex,
+        _dynamic_groups,
+        _dynamic_matches,
+        _dynamic_built_route
+      ) =>
+        // Check if the route requires a data file
+        s.has(route) &&
+        (isDynamicRoute(route)
+          ? ((_dynamic_regex = getRouteRegex(route)),
+            (_dynamic_groups = _dynamic_regex.groups),
+            (_dynamic_matches =
+              getRouteMatcher(_dynamic_regex)(asPathname) || query),
+            (_dynamic_built_route = route),
+            Object.keys(_dynamic_groups).every(
+              param =>
+                param in _dynamic_matches &&
+                // Interpolate group into data URL if present
+                (_dynamic_built_route = _dynamic_built_route.replace(
+                  `[${_dynamic_groups[param].repeat ? '...' : ''}${param}]`,
+                  _dynamic_groups[param].repeat
+                    ? // FIXME: assuming `_matches` is an array
+                      _dynamic_matches[param].map(encodeURIComponent).join('/')
+                    : encodeURIComponent(_dynamic_matches[param])
+                ))
+            ) &&
+              (_dataHref = `${this.assetPrefix}/_next/data/${this.buildId}${_dynamic_built_route}.json`))
```

Comment:
would it reduce the size any to make a util to build this href so we can use it here and the next line and for the current prefetching on line 303?

Reply:
Extracted helper